beam/src/main/scala/magnolify/beam/BeamSchemaType.scala

```diff
@@ -0,0 +1,283 @@
+        val idx = v.getCaseType.getValue
+        sealedTrait.subtypes.find(_.index == idx) match {
+          case None      => throw new IllegalArgumentException(s"OneOf index not found: [$idx]")
+          case Some(sub) => sub.typeclass.fromAny(v.getValue)(cm)
+        }
+      }
+
+      def to(v: T)(cm: CaseMapper): logicaltypes.OneOfType.Value =
+        sealedTrait.split(v)(sub =>
+          beamOneOfType(cm).createValue(enumName(sub), sub.typeclass.to(sub.cast(v))(cm))
+        )
+    }
+
+  def join[T](caseClass: CaseClass[Typeclass, T]): BeamSchemaField[T] = {
+    if (caseClass.isValueClass) {
+      val p = caseClass.parameters.head
+      val tc = p.typeclass
+      new BeamSchemaField[T] {
+        override type FromT = tc.FromT
+        override type ToT = tc.ToT
+        override def fieldType(cm: CaseMapper): FieldType = tc.fieldType(cm)
+        override def from(v: FromT)(cm: CaseMapper): T = caseClass.construct(_ => tc.fromAny(v)(cm))
+        override def to(v: T)(cm: CaseMapper): ToT = tc.to(p.dereference(v))(cm)
+      }
+    } else {
+      new Record[T] {
+        override def fieldType(cm: CaseMapper): FieldType = FieldType.row(schema(cm))
+
+        override protected def buildSchema(cm: CaseMapper): Schema =
+          caseClass.parameters
+            .foldLeft(Schema.builder()) { case (s, p) =>
+              s.addField(cm.map(p.label), p.typeclass.fieldType(cm))
+            }
+            .build()
+
+        override def from(v: Row)(cm: CaseMapper): T =
+          caseClass.construct(p => p.typeclass.fromAny(v.getValue[Any](p.index))(cm))
+
+        override def to(v: T)(cm: CaseMapper): Row = {
+          val values = caseClass.parameters.map { p =>
+            p.typeclass.to(p.dereference(v))(cm).asInstanceOf[Object]
+          }
+          Row.withSchema(schema(cm)).addValues(values.asJava).build()
+        }
+      }
+    }
+  }
+
+  // BYTE	An 8-bit signed value
+  implicit val bsfByte: BeamSchemaField[Byte] = id[Byte](_ => FieldType.BYTE)
+  // INT16	A 16-bit signed value
+  implicit val bsfShort: BeamSchemaField[Short] = id[Short](_ => FieldType.INT16)
+  implicit val bsfChar: BeamSchemaField[Char] = from[Short](_.toChar)(_.toShort)
+  // INT32	A 32-bit signed value
+  implicit val bsfInt: BeamSchemaField[Int] = id[Int](_ => FieldType.INT32)
+  // INT64	A 64-bit signed value
+  implicit val bsfLong: BeamSchemaField[Long] = id[Long](_ => FieldType.INT64)
+  // FLOAT	A 32-bit IEEE 754 floating point number
+  implicit val bsfFloat: BeamSchemaField[Float] = id[Float](_ => FieldType.FLOAT)
+  // DOUBLE	A 64-bit IEEE 754 floating point number
+  implicit val bsfDouble: BeamSchemaField[Double] = id[Double](_ => FieldType.DOUBLE)
+  // STRING	A string
+  implicit val bsfString: BeamSchemaField[String] = id[String](_ => FieldType.STRING)
+  implicit val bsfCharSeq: BeamSchemaField[CharSequence] =
+    from[String](_.asInstanceOf[CharSequence])(_.toString)
```

Comment:
Do we want this one? I think only avro has this type-class for historical reason